src/backend/access/appendonly/aosegfiles.c

```
@@ -2175,117 +2175,89 @@ PrintPgaosegAndGprelationNodeEntries(FileSegInfo **allseginfo, int totalsegs, bo
-		}
-
-		if (segmentCount > totalsegs + 1)
-		{
-			PrintPgaosegAndGprelationNodeEntries(allseginfo, totalsegs, segmentFileNumMap);
-			elog(ERROR, "gp_relation_node (%d) has more entries than pg_aoseg (%d) for relation %s",
-				 segmentCount,
-				 totalsegs,
-				 RelationGetRelationName(rel));
-		}
+		if (gprelnodeMap[segmentFileNum] != true)
+			gprelnodeMap[segmentFileNum] = true;
 	}
-
 	GpRelationNodeEndScan(&gpRelationNodeScan);
 	heap_close(gp_relation_node, AccessShareLock);
 
-	if (totalsegs > 0)
-	{
-		if (allseginfo[0]->segno == 0)
-		{
-			i++;
-		}
-	}
-
-	for (j = 1; j < num_gp_relation_node_entries; j++)
+	/*
+	 * Build aosegMap. It is also indexed by segno.
+	 */
+	for (i = 0; i < totalsegs; i++)
 	{
-		if (segmentFileNumMap[j] == true)
-		{
-			while (i < totalsegs && allseginfo[i]->segno != j)
-			{
+		FileSegInfo *fsinfo = allseginfo[i];
 
-				if (allseginfo[i]->eof != 0 && segmentFileNumMap[allseginfo[i]->segno] == false)
-				{
-					PrintPgaosegAndGprelationNodeEntries(allseginfo, totalsegs, segmentFileNumMap);
-					elog(ERROR, "Missing pg_aoseg entry %d in gp_relation_node for %s",
-						 allseginfo[i]->segno, RelationGetRelationName(rel));
-				}
-				i++;
-			}
-
-			if (i < totalsegs && allseginfo[i]->segno == j)
-			{
-				i++;
-				continue;
-			}
-
-			if (i == totalsegs)
-			{
-				PrintPgaosegAndGprelationNodeEntries(allseginfo, totalsegs, segmentFileNumMap);
-				elog(ERROR, "Missing gp_relation_node entry %d in pg_aoseg for %s",
-					 j, RelationGetRelationName(rel));
-			}
-		}
+		if (fsinfo->segno < 0 || fsinfo->segno > AOTupleId_MaxSegmentFileNum)
+			elog(ERROR, "invalid segno on pg_aoseg row: %d", fsinfo->segno);
+		aosegMap[fsinfo->segno] = fsinfo;
 	}
 
 	/*
-	 * Check for any extra entries in pg_aoseg which are not present in
-	 * gp_relation_node
+	 * Check that there is a one-to-one correspondence between the two maps.
 	 */
-	while (i < totalsegs)
+	for (i = 0; i <= AOTupleId_MaxSegmentFileNum; i++)
```

Comes with cost of looping for `AOTupleId_MaxSegmentFileNum` even if table has only 1 segfile, can we add fail fast logic please.